kv/kvserver: TestReplicateReAddAfterDown failed

[(kv/kvserver).TestReplicateReAddAfterDown failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2618017&tab=buildLog) on [master@081f813ae4fb04653909d2eebd862e5965e21878](https://github.com/cockroachdb/cockroach/commits/081f813ae4fb04653909d2eebd862e5965e21878):

```
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:331
          | github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:863
          | github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:901
          | github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.Batch
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:465
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).sendBatch
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:193
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).SendNext
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:175
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendToReplicas
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1884
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1504
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1142
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:781
          | github.com/cockroachdb/cockroach/pkg/kv.(*CrossRangeTxnWrapperSender).Send
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:219
          | github.com/cockroachdb/cockroach/pkg/kv.(*DB).sendUsingSender
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:794
          | github.com/cockroachdb/cockroach/pkg/kv.(*DB).send
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:777
          | github.com/cockroachdb/cockroach/pkg/kv.sendAndFill
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:705
          | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Run
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:728
          | github.com/cockroachdb/cockroach/pkg/kv.(*DB).AdminChangeReplicas
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:633
          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).changeReplicas
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:588
          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).addReplica
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:602
          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).AddVoters
          | 	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:620
        Wraps: (5) failed to connect to n3 at 127.0.0.1:39511
        Wraps: (6) initial connection heartbeat failed
          |
          | (opaque error wrapper)
          | type name: github.com/cockroachdb/cockroach/pkg/util/netutil/*netutil.InitialHeartbeatFailedError
        Wraps: (7) rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:39511: connect: connection refused"
          |
          | (opaque error leaf)
          | type name: google.golang.org/grpc/internal/status/*status.Error
        Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *markers.withMark (4) *errbase.opaqueWrapper (5) *errutil.withPrefix (6) *errbase.opaqueWrapper (7) *errbase.opaqueLeaf
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicateReAddAfterDown631307767
--- FAIL: TestReplicateReAddAfterDown (0.86s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestReplicateReAddAfterDown PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestReplicateReAddAfterDown.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

@tbg here is the full error message, from another repro:
`I210127 17:19:03.772933 1 (gostd) rand.go:92  [-] 1  random seed: -132423090394599937
=== RUN   TestReplicateReAddAfterDown
    test_log_scope.go:73: test logs captured to: /var/folders/5x/_zb1l2tj7xnbq3vyq79zcfrc0000gn/T/logTestReplicateReAddAfterDown127474313
    test_log_scope.go:74: use -show-logs to present logs inline
    client_raft_test.go:3770: could not add [n3,s3] replicas to range containing "a": AdminChangeReplicas error: failed to connect to n3 at 127.0.0.1:56528: initial connection heartbeat failed: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:56528: connect: connection refused"
        (1) attached stack trace
          -- stack trace:
          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).changeReplicas
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:592
          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).addReplica
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:602
          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).AddVoters
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:620
          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).AddVotersOrFatal
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:728
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicateReAddAfterDown
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_raft_test.go:3770
          | testing.tRunner
          | 	/usr/local/opt/go/libexec/src/testing/testing.go:1123
          | runtime.goexit
          | 	/usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1374
        Wraps: (2) AdminChangeReplicas error
        Wraps: (3) forced error mark
          | "snapshot failed"
          | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        Wraps: (4)
          | (opaque error wrapper)
          | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
          | reportable 0:
          |
          | github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).dial
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:169
          | github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).Dial
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:90
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*RaftTransport).SendSnapshot
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/raft_transport.go:656
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendSnapshot
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_command.go:2112
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).atomicReplicationChange
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_command.go:1418
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).changeReplicasImpl
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_command.go:1045
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).ChangeReplicas
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_command.go:964
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeAdminBatch
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:561
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithRangeID
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:99
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:35
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:195
          | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).Send
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores.go:177
          | github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal.func1
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:875
          | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:331
          | github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:863
          | github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:901
          | github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.Batch
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:465
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).sendBatch
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:193
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).SendNext
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:175
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendToReplicas
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1872
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1503
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1141
          | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:780
          | github.com/cockroachdb/cockroach/pkg/kv.(*CrossRangeTxnWrapperSender).Send
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:219
          | github.com/cockroachdb/cockroach/pkg/kv.(*DB).sendUsingSender
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:794
          | github.com/cockroachdb/cockroach/pkg/kv.(*DB).send
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:777
          | github.com/cockroachdb/cockroach/pkg/kv.sendAndFill
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:705
          | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Run
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:728
          | github.com/cockroachdb/cockroach/pkg/kv.(*DB).AdminChangeReplicas
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:633
          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).changeReplicas
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:588
          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).addReplica
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:602
          | github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).AddVoters
          | 	/Users/alunev/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:620
        Wraps: (5) failed to connect to n3 at 127.0.0.1:56528
        Wraps: (6) initial connection heartbeat failed
          |
          | (opaque error wrapper)
          | type name: github.com/cockroachdb/cockroach/pkg/util/netutil/*netutil.InitialHeartbeatFailedError
        Wraps: (7) rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:56528: connect: connection refused"
          |
          | (opaque error leaf)
          | type name: google.golang.org/grpc/internal/status/*status.Error
        Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *markers.withMark (4) *errbase.opaqueWrapper (5) *errutil.withPrefix (6) *errbase.opaqueWrapper (7) *errbase.opaqueLeaf
    panic.go:617: -- test log scope end --
test logs left over in: /var/folders/5x/_zb1l2tj7xnbq3vyq79zcfrc0000gn/T/logTestReplicateReAddAfterDown127474313
--- FAIL: TestReplicateReAddAfterDown (1.19s)
FAIL`

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -3711,7 +3711,6 @@ func TestReplicaLazyLoad(t *testing.T) {
 
 func TestReplicateReAddAfterDown(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 59453, "flaky test")
 	defer log.Scope(t).Close(t)
 
 	stickyEngineRegistry := server.NewStickyInMemEnginesRegistry()
@@ -5124,9 +5123,9 @@ func TestProcessSplitAfterRightHandSideHasBeenRemoved(t *testing.T) {
 
 		// Restart store 0 so that it forgets about the newer replicaID.
 		tc.StopServer(0)
+		lhsPartition.deactivate()
 		require.NoError(t, tc.RestartServer(0))
 
-		lhsPartition.deactivate()
 		tc.WaitForValues(t, keyA, []int64{8, 8, 8})
 		hs := getHardState(t, tc.GetFirstStoreFromServer(t, 0), rhsID)
 		require.Equal(t, uint64(0), hs.Commit)
@@ -5197,7 +5196,7 @@ func TestProcessSplitAfterRightHandSideHasBeenRemoved(t *testing.T) {
 	})
 
 	// This case is set up like the previous one except after the RHS learns about
-	// its higher replica ID the store crahes and forgets. The RHS replica gets
+	// its higher replica ID the store crashes and forgets. The RHS replica gets
 	// initialized by the split.
 	t.Run("(4) initial replica RHS partition, with restart", func(t *testing.T) {
 		tc, db, keyA, keyB, _, lhsPartition, stickyEngineRegistry := setup(t)
@@ -5252,9 +5251,9 @@ func TestProcessSplitAfterRightHandSideHasBeenRemoved(t *testing.T) {
 		}
 
 		tc.StopServer(0)
+		lhsPartition.deactivate()
 		require.NoError(t, tc.RestartServer(0))
 
-		lhsPartition.deactivate()
 		tc.WaitForValues(t, keyA, []int64{8, 8, 8})
 		// In this case the store has forgotten that it knew the RHS of the split
 		// could not exist. We ensure that it has been initialized to the initial

--- a/pkg/testutils/testcluster/BUILD.bazel
+++ b/pkg/testutils/testcluster/BUILD.bazel
@@ -19,6 +19,7 @@ go_library(
         "//pkg/storage",
         "//pkg/testutils",
         "//pkg/testutils/serverutils",
+        "//pkg/util/contextutil",
         "//pkg/util/hlc",
         "//pkg/util/log",
         "//pkg/util/retry",

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -33,6 +33,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/storage"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/util/contextutil"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/retry"
@@ -1213,7 +1214,7 @@ func (tc *TestCluster) RestartServer(idx int) error {
 // RestartServerWithInspect uses the cached ServerArgs to restart a Server
 // specified by the passed index. We allow an optional inspect function to be
 // passed in that can observe the server once its been re-created but before it's
-// bean started. This is useful for tests that want to capture that the startup
+// been started. This is useful for tests that want to capture that the startup
 // sequence performs the correct actions i.e. that on startup liveness is gossiped.
 func (tc *TestCluster) RestartServerWithInspect(idx int, inspect func(s *server.TestServer)) error {
 	if !tc.ServerStopped(idx) {
@@ -1251,26 +1252,68 @@ func (tc *TestCluster) RestartServerWithInspect(idx int, inspect func(s *server.
 	}
 	s := srv.(*server.TestServer)
 
-	tc.mu.Lock()
-	defer tc.mu.Unlock()
-	tc.Servers[idx] = s
-	tc.mu.serverStoppers[idx] = s.Stopper()
+	if err := func() error {
+		tc.mu.Lock()
+		defer tc.mu.Unlock()
+		tc.Servers[idx] = s
+		tc.mu.serverStoppers[idx] = s.Stopper()
 
-	if inspect != nil {
-		inspect(s)
-	}
+		if inspect != nil {
+			inspect(s)
+		}
 
-	if err := srv.Start(); err != nil {
+		if err := srv.Start(); err != nil {
+			return err
+		}
+
+		dbConn, err := serverutils.OpenDBConnE(srv.ServingSQLAddr(),
+			serverArgs.UseDatabase, serverArgs.Insecure, srv.Stopper())
+		if err != nil {
+			return err
+		}
+		tc.Conns[idx] = dbConn
+		return nil
+	}(); err != nil {
 		return err
 	}
 
-	dbConn, err := serverutils.OpenDBConnE(srv.ServingSQLAddr(),
-		serverArgs.UseDatabase, serverArgs.Insecure, srv.Stopper())
-	if err != nil {
-		return err
-	}
-	tc.Conns[idx] = dbConn
-	return nil
+	// Wait until the other nodes can successfully connect to the newly restarted
+	// node. This is useful to avoid flakes: the newly restarted node is now on a
+	// different port, and a cycle of gossip is necessary to make all other nodes
+	// aware.
+	return contextutil.RunWithTimeout(
+		context.Background(), "check-conn", 15*time.Second,
+		func(ctx context.Context) error {
+			r := retry.StartWithCtx(ctx, retry.Options{
+				InitialBackoff: 1 * time.Millisecond,
+				MaxBackoff:     100 * time.Millisecond,
+			})
+			var err error
+			for r.Next() {
+				err = func() error {
+					for idx, s := range tc.Servers {
+						if tc.ServerStopped(idx) {
+							continue
+						}
+						for i := 0; i < rpc.NumConnectionClasses; i++ {
+							class := rpc.ConnectionClass(i)
+							if _, err := s.NodeDialer().Dial(ctx, srv.NodeID(), class); err != nil {
+								return errors.Wrapf(err, "connecting n%d->n%d (class %v)", s.NodeID(), srv.NodeID(), class)
+							}
+						}
+					}
+					return nil
+				}()
+				if err != nil {
+					continue
+				}
+				return nil
+			}
+			if err != nil {
+				return err
+			}
+			return ctx.Err()
+		})
 }
 
 // ServerStopped determines if a server has been explicitly

